File: src/main/java/offerII/Offer2.java
package offerII;

/**
 * @author jingxinwu
 * @date 2021-12-06 11:44 下午
 */
public class Offer2 {

    public static void main(String[] args) {
        System.out.println(new Offer2().addBinary("11", "10"));
    }

    public String addBinary(String a, String b) {
        StringBuilder ans = new StringBuilder();
        int n = Math.max(a.length(), b.length()), carry = 0;
        for (int i = 0; i < n; ++i) {
            carry += i < a.length() ? (a.charAt(a.length() - 1 - i) - '0') : 0;
            carry += i < b.length() ? (b.charAt(b.length() - 1 - i) - '0') : 0;
            ans.append((char) (carry % 2 + '0'));
            carry /= 2;
        }

        if (carry > 0) {
            ans.append('1');
        }
        ans.reverse();
        return ans.toString();
    }
}

File: src/main/GO/leetcode/lists/offerII/Offer2.go
package offerII

import "strconv"

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/12 14:21
 */

func addBinary(a string, b string) string {
	carry := 0
	res := ""
	i, j := len(a)-1, len(b)-1
	for i >= 0 || j >= 0 {
		var one, two int
		if i < 0 {
			one = 0
		} else {
			one = int(a[i] - '0')
		}
		if j < 0 {
			two = 0
		} else {
			two = int(b[j] - '0')
		}
		t := one ^ two ^ carry
		if one == 1 && two == 1 {
			carry = 1
		} else {
			carry = (one ^ two) & carry
		}
		res = strconv.Itoa(t) + res
		i--
		j--
	}
	if carry == 1 {
		return strconv.Itoa(carry) + res
	}
	return res
}
